Binarise

HedvigS opened the pull request:
Function that binarises grambank data correctly from a cldf values table

johenglisch commented:
Some remarks on the test code:

 1. Put your test files into `tests/testthat`.
 2. Name them something like `tests/testthat/test-binarise.R`.
 3. Structure the test code in these little blocks:

    ```
    test_that("multistate features are being binarised", {
      data_with_multistate_features <- ...
      outcome <- binarise(data_with_multistate_features) %>% ...
      expected <- ...
      expect_equal(outcome, expected)
    })

    test_that("binarised features are not being overwritten", {
      data_with_binarised_features <- ...
      outcome <- binarise(data_with_binarised_features) %>% ...
      expected <- ...
      expect_equal(outcome, expected)
    })
    ```

([like this](https://github.com/grambank/rgrambank/blob/9721e30614745504fc52e8d0ccd5be93c8559e42/tests/testthat/test-db.R), basically)

If you do that you can just hit a key combination in RStudio ([Cmd+Shift+T, apparently](https://testthat.r-lib.org/)) and it will automatically find all your tests and run them.  You'll get a nice summary in the *Build* tab:

    ℹ Testing rgrambank
    [...]
    ✔ | F W S  OK | Context
    ✔ |         2 | db

    ══ Results ═════════════════════════════════════════════
    [ FAIL 0 | WARN 0 | SKIP 0 | PASS 2 ]

Advantages:

 1. You can get into a really nice workflow like *‘Change code – run tests – fix code – run tests – add test – run tests – fix code – run tests’* etc.

 2. Github can be set up to run tests automatically every time someone pushes to master or submits a PR, which is reeeally nice ([we do that for our python programs](https://github.com/grambank/pygrambank/blob/master/.github/workflows/python-package.yml) and I assume it's probably possible with R as well).

Also, it's good practice to have your tests be as isolated as possible from the ‘messy  real world’, i.e. not relying on a working internet connection or a remote file not being deleted:

https://github.com/grambank/rgrambank/blob/7a43c78ede5b36d42a65ff0f89dba34649b76e08/tests/test_binarise.R#L10

It might be better to read example data from a local file in the git repo or to manually create the data in the test program itself, like what you did with `fake_raw_binary_data` below (that's usually what I prefer):

https://github.com/grambank/rgrambank/blob/7a43c78ede5b36d42a65ff0f89dba34649b76e08/tests/test_binarise.R#L12-L22

That will make it so your tests only fail (or stop failling) because of *changes to your code*, not because of outside factors like working on a train ride.

HedvigS commented:
> Some remarks on the test code:
> 
> 1. Put your test files into `tests/testthat`.
> 2. Name them something like `tests/testthat/test-binarise.R`.
> 3. Structure the test code in these little blocks:
>    ```
>    test_that("multistate features are being binarised", {
>      data_with_multistate_features <- ...
>      outcome <- binarise(data_with_multistate_features) %>% ...
>      expected <- ...
>      expect_equal(outcome, expected)
>    })
>    
>    test_that("binarised features are not being overwritten", {
>      data_with_binarised_features <- ...
>      outcome <- binarise(data_with_binarised_features) %>% ...
>      expected <- ...
>      expect_equal(outcome, expected)
>    })
>    ```
> 
> ([like this](https://github.com/grambank/rgrambank/blob/9721e30614745504fc52e8d0ccd5be93c8559e42/tests/testthat/test-db.R), basically)
> 
> If you do that you can just hit a key combination in RStudio ([Cmd+Shift+T, apparently](https://testthat.r-lib.org/)) and it will automatically find all your tests and run them. You'll get a nice summary in the _Build_ tab:
> 
> ```
> ℹ Testing rgrambank
> [...]
> ✔ | F W S  OK | Context
> ✔ |         2 | db
> 
> ══ Results ═════════════════════════════════════════════
> [ FAIL 0 | WARN 0 | SKIP 0 | PASS 2 ]
> ```
> 
> Advantages:
> 
> 1. You can get into a really nice workflow like _‘Change code – run tests – fix code – run tests – add test – run tests – fix code – run tests’_ etc.
> 2. Github can be set up to run tests automatically every time someone pushes to master or submits a PR, which is reeeally nice ([we do that for our python programs](https://github.com/grambank/pygrambank/blob/master/.github/workflows/python-package.yml) and I assume it's probably possible with R as well).
> 
> Also, it's good practice to have your tests be as isolated as possible from the ‘messy real world’, i.e. not relying on a working internet connection or a remote file not being deleted:
> 
> https://github.com/grambank/rgrambank/blob/7a43c78ede5b36d42a65ff0f89dba34649b76e08/tests/test_binarise.R#L10
> 
> It might be better to read example data from a local file in the git repo or to manually create the data in the test program itself, like what you did with `fake_raw_binary_data` below (that's usually what I prefer):
> 
> https://github.com/grambank/rgrambank/blob/7a43c78ede5b36d42a65ff0f89dba34649b76e08/tests/test_binarise.R#L12-L22
> 
> That will make it so your tests only fail (or stop failling) because of _changes to your code_, not because of outside factors like working on a train ride.

Sure, I can do that if we want that. I just wrote these scripts quickly for myself without testthat to see that the code worked the way I expected.

@SimonGreenhill do you want all functions to have a formalised testthat workflow as Johannes suggests?

HedvigS commented:
@johenglisch does the binarise-function work the way you expect it too based on our chat the other day?

SimonGreenhill commented:
yes please! it'll make life better in the long run

HedvigS commented:
> yes please! it'll make life better in the long run

Alright, I'll try and find time later this week to do this.

HedvigS commented:
I'm working in this branch now. i can't turn this PR into a draft, but if I could I would and turn it back when I'm done :)

HedvigS commented:
alright @johenglisch and @SimonGreenhill I've made a stab at a testthat for binarise.

johenglisch commented:
The tests look good.  Just a tiny hint:  You don't need to put the `library(...)` calls inside the individual test cases.

Like, as a general rule it's very good that the tests don't share any variables etc. among each other.  We like every test to fail/succeed on its own, independent of other test code.  But there are things that are considered ‘fair game’ for tests to share (esp. library imports and the occasional constant):

    library(...)
    library(...)

    test_that("...", {
      ...
    })

    test_that("...", {
      ...
    })

    test_that("...", {
      ...
    })

It just makes writing/changing tests a bit less tedious for you.

HedvigS commented:
> The tests look good. Just a tiny hint: You don't need to put the `library(...)` calls inside the individual test cases.
> 
> Like, as a general rule it's very good that the tests don't share any variables etc. among each other. We like every test to fail/succeed on its own, independent of other test code. But there are things that are considered ‘fair game’ for tests to share (esp. library imports and the occasional constant):
> 
> ```
> library(...)
> library(...)
> 
> test_that("...", {
>   ...
> })
> 
> test_that("...", {
>   ...
> })
> 
> test_that("...", {
>   ...
> })
> ```
> 
> It just makes writing/changing tests a bit less tedious for you.

Thanks! I had those in because they were necessary when I was trial and erroring at one stage, but of course happy to remove now.

HedvigS commented:
> @johenglisch does the binarise-function work the way you expect it too based on our chat the other day?

@johenglisch besides the tests, does the function work the way you expect?

johenglisch commented:
>does the function work the way you expect?

These tests here look exactly like your notes on the whiteboard, so I'd say it's working:

https://github.com/grambank/rgrambank/blob/653f16b9127aa475255260133fe9ac532db50fb0/tests/testthat/test-binarise.R#L3-L9

HedvigS commented:
@SimonGreenhill I'm gonna go ahead and re-write binarise now as per your email

HedvigS commented on R/as.grambank.wide.R:
```diff
@@ -0,0 +1,20 @@
+#' Extract a wide dataframe from a grambank value table
```
I think that a function like this could live in rcldf instead. I get questions more often than I'd expect on how to make ValueTables wide, and this function and the one I wrote before are super simple and works on any cldf ValueTable. I think people just need to be hand-held through reshape::dcast/tidyr::pivot_wider/tidyr::spread sometimes

xrotwang replied:
A function like that in `rcldf` would need to identify relevant columns by propertyUrl in the associated metadata, though, not by column name.

SimonGreenhill replied:
yes, the implementation in rcldf would need to be more complicated -- and for some datasets could be explosively bad (e.g. dplace has 2500 variables, do people really want a 2500 col dataframe?) I think we should keep this here for now for grambank purposes and think about implementing something in rcldf later.

HedvigS commented:
i removed the need for stringr

HedvigS commented on R/binarise.R:
```diff
@@ -0,0 +1,253 @@
+    wide_values$GB024a <- binarise_gb024_to_gb024a(wide_values$GB024)
+    wide_values$GB024b <- binarise_gb024_to_gb024b(wide_values$GB024)
+
+    # are we deliberately using dplyr::if_else AND base::ifelse?
```
@SimonGreenhill yes we were in grambank-analysed because of reasons to do with NAs.
I was told to not spend too much time perfecting that codebase but just make it work, so I made changes to make it work but the overall code wasn't always as elegant as it could be.

HedvigS replied:
This comment was in relation to simon asking about the combo of ifelse and if_else. It's no longer relevant, resolving now

HedvigS commented:
@SimonGreenhill I'm done with `tests/testthat/test-binarise.R` and `R/binarise.R`.

Turning to language_level_df now

HedvigS commented:
@SimonGreenhill I suspect something unfortunate has happened, because I made changes to language_level_df to take only long tables this week, but I think what got merged into here doesn't reflect that. i'm gonna do some git detective work and double check things.

HedvigS commented:
> @SimonGreenhill I suspect something unfortunate has happened, because I made changes to language_level_df to take only long tables this week, but I think what got merged into here doesn't reflect that. i'm gonna do some git detective work and double check things.

Yep, I can confirm the language_level_df script in this branch is older than in the branch in #3 . I'll go ahead and do the specific pull ins.

SimonGreenhill commented:
Don't think I changed anything vital in that file beyond adding comments so you should be ok to just overwrite

HedvigS commented:
@SimonGreenhill I have moved content over from #3 so that the function language_level_df again operates with long tables rather than wide. I think I've addressed all of your feedback on that function now and I made some other smaller improvements.

HedvigS commented:
@SimonGreenhill  binarise and language_level_df are ready for review again.

xrotwang commented on R/binarise.R:
```diff
@@ -0,0 +1,119 @@
+
+# BINARISING MULTISTATE FEATUYES
+
+    ValueTable <- gb_recode(ValueTable, 'GB024', 'GB024a', binarise_GBXXX_to_GBXXXa_without_zero)
```
Looking up the literal strings in `binary_parameters`/`multistate_parameters` would be more elegant and less prone to typos.

HedvigS replied:
sure. Ehm @SimonGreenhill do you want that?

HedvigS replied:
NB, I've split the set in two groups depending on wether they can take a 0 value or not. Perhaps a bit extra conservative, but I like sanity checks.

xrotwang replied:
Maybe a config dict like
```python
multistate_parameters = {
    'GB024': False,  # without zero
    ...
    'GB203': True,  # with zero
}
```
would do, then.

It just looked strange that a list of the literal feature IDs was defined above, and then not used down here.

xrotwang replied:
```python
for fid, with_zero in multistate_parameters.items():
    for ms in 'ab':
        ValueTable = gb_recode(ValueTable, fid, fid + ms, binarise_*_with_zero if with_zero else binarise_*_without_zero)
```

HedvigS replied:
@xrotwang I see what you mean. i use it in another place
@SimonGreenhill decides

SimonGreenhill replied:
yes, agree that it's more elegant. I'll merge this now and make that a new issue (#16) for later refactoring.

xrotwang commented on R/language_level_df.R:
```diff
@@ -0,0 +1,145 @@
+}
+
+# The language level ID column is named different in different, CLDF datasets.
+# setting them to the same
```
I don't really understand this bit. AFAICT the function expects the LanguageTable from the glottolog-cldf dataset. Which other CLDF datasets with different column names would make sense as input here?

HedvigS replied:
It can also take a LanguageTable from some other cldf-dataset. 

This function can work on several different kinds of cldf-datasets where language-levelling is relevant. I've suggested before it could go in rcldf instead of rgrambank

HedvigS replied:
I set it to read glottolog 4.8 from github just as a default fallback.

xrotwang replied:
Ah, ok, I see that the relevant information is also available from the Grambank LanguageTable (to which it was copied from Glottolog for convenience). But different "levels" of languoids and relations between these via `Language_level_ID` aren't standardized in CLDF, but just Glottolog-specific ways to encode dialect/language classifications.

So I'd either require Glottolog's LanguageTable to be passed into this function, or just have one dataset (or `Grambank`) argument, and then use ValueTable and LanguageTable from the same dataset in this function.

HedvigS replied:
I think the solution here depends on if this function continues to live in rgrambank or is moved to rcldf.

SimonGreenhill replied:
I agree with robert here -- this function needs to know three things:
1. a language table from a cldf dataset
2. a values table from a cldf dataset.
3. a glottolog dataset.

The current implementation relies on grambank only columns (Language_level_ID) which is ok, but it would be nice to make it more general (which isn't too far off but will require pulling this out of glottolog). 

More problematically,  it relies on a hardcoded table from glottolog. We should decouple this because glottolog evolves, and we don't want to be updating this every single time glottolog releases a new version. The package here has  a function to easily load_glottolog() so we can make users use that e.g. 

```r
df <- language_level_df(df, glottolog=load_glottolog())
```

I'll merge this for now and fix this myself.

SimonGreenhill replied:
c.f. #17
